wolfcrypt/src/rc2.c

```diff
@@ -0,0 +1,335 @@
+        j = i * 4;
+
+        /* mixing round */
+        r10 = r10 + key[j] + (r76 & r54) + (~r76 & r32);
+        r10 = rotlFixed16(r10, 1);
+
+        r32 = r32 + key[j+1] + (r10 & r76) + (~r10 & r54);
+        r32 = rotlFixed16(r32, 2);
+
+        r54 = r54 + key[j+2] + (r32 & r10) + (~r32 & r76);
+        r54 = rotlFixed16(r54, 3);
+
+        r76 = r76 + key[j+3] + (r54 & r32) + (~r54 & r10);
+        r76 = rotlFixed16(r76, 5);
+
+        /* mashing round on loop 5, 11 */
+        if (i == 4 || i == 10) {
+            r10 = r10 + key[r76 & 63];
+            r32 = r32 + key[r10 & 63];
+            r54 = r54 + key[r32 & 63];
+            r76 = r76 + key[r54 & 63];
+        }
+    }
+
+    out[0] = (byte)r10;
+    out[1] = (byte)(r10 >> 8);
+    out[2] = (byte)r32;
+    out[3] = (byte)(r32 >> 8);
+    out[4] = (byte)r54;
+    out[5] = (byte)(r54 >> 8);
+    out[6] = (byte)r76;
+    out[7] = (byte)(r76 >> 8);
+
+    return 0;
+}
+
+/**
+   RC2 ECB decrypt operation on one single RC2_BLOCK_SIZE block.
+   rc2      Initialized RC2 structure
+   out      [out] Destination for decrypted plaintext
+   in       Input ciphertext to be decrypted
+   sz       Size of the output buffer, out
+   return   0 on success, negative on error
+*/
+int wc_Rc2EcbDecrypt(Rc2* rc2, byte* out, const byte* in, word32 sz)
+{
+    int i, j = 63;
+    word16 r0, r1, r2, r3;
+    word16* key;
+
+    if (rc2 == NULL || out == NULL || in == NULL) {
+        return BAD_FUNC_ARG;
+    }
+    key = rc2->key;
+
+    if (sz != RC2_BLOCK_SIZE) {
+        return BUFFER_E;
+    }
+
+    r0 = (in[1] << 8) | in[0];
+    r1 = (in[3] << 8) | in[2];
+    r2 = (in[5] << 8) | in[4];
+    r3 = (in[7] << 8) | in[6];
+
+    for (i = 16; i > 0; i--) {
+        j = 4*i - 1;
+
+        r3 = rotrFixed16(r3, 5);
+        r3 = r3 - key[j] - (r2 & r1) - (~r2 & r0);
+
+        r2 = rotrFixed16(r2, 3);
+        r2 = r2 - key[j-1] - (r1 & r0) - (~r1 & r3);
+
+        r1 = rotrFixed16(r1, 2);
+        r1 = r1 - key[j-2] - (r0 & r3) - (~r0 & r2);
+
+        r0 = rotrFixed16(r0, 1);
+        r0 = r0 - key[j-3] - (r3 & r2) - (~r3 & r1);
+
+        if (i == 12 || i == 6) {
+            r3 = r3 - key[r2 & 63];
+            r2 = r2 - key[r1 & 63];
+            r1 = r1 - key[r0 & 63];
+            r0 = r0 - key[r3 & 63];
+        }
+    }
+
+    out[0] = (byte)r0;
+    out[1] = (byte)(r0 >> 8);
+    out[2] = (byte)r1;
+    out[3] = (byte)(r1 >> 8);
+    out[4] = (byte)r2;
+    out[5] = (byte)(r2 >> 8);
+    out[6] = (byte)r3;
+    out[7] = (byte)(r3 >> 8);
+
+    return 0;
+}
+
+int wc_Rc2CbcEncrypt(Rc2* rc2, byte* out, const byte* in, word32 sz)
+{
+    word32 blocks = (sz / RC2_BLOCK_SIZE);
+
+    if (rc2 == NULL || out == NULL || in == NULL) {
+        return BAD_FUNC_ARG;
+    }
+
+    if (sz == 0) {
+        return 0;
+    }
+
+    while (blocks--) {
+        xorbuf((byte*)rc2->reg, in, RC2_BLOCK_SIZE);
+        wc_Rc2EcbEncrypt(rc2, (byte*)rc2->reg, (byte*)rc2->reg, RC2_BLOCK_SIZE);
```

Comment:
Return code check

Reply:
Done, thanks.  Had been following the precedence of other CBC mode ops.